king_phisher/client/windows/campaign_import.py

```diff
@@ -0,0 +1,439 @@
+		"""
+		Used by the import thread to import the campaign into the database.
+		Through this process after every major process, the thread will check to see if it has been stopped.
+		"""
+		if not self.campaign_info:
+			return
+		# prevent user from changing campaign info during import
+		start_time = datetime.datetime.now()
+		GLib.idle_add(self.button_import_campaign.set_sensitive, False)
+		GLib.idle_add(self.select_button.set_sensitive, False)
+		GLib.idle_add(self.spinner.start)
+
+		batch_size = 100
+		if self.thread_import_campaign.stopped():
+			return
+		self.prepep_xml_data()
+
+		self.campaign_info.find('id').text = self.rpc(
+			'campaign/new',
+			self.campaign_info.find('name').text,
+			self.campaign_info.find('description').text
+		)
+		self.logger.info('created new campaign id: {}'.format(self.campaign_info.find('id').text))
+
+		nodes_completed = 0
+		node_count = float(len(self.campaign_info.findall('.//*')))
+		if self.thread_import_campaign.stopped():
+			return
+		for nods in self.campaign_info.getiterator():
+			if nods.tag == 'campaign_id':
+				nods.text = self.campaign_info.find('id').text
+		self._update_text_view('Campaign created, ID set to {}'.format(self.campaign_info.find('id').text), idle=True)
+
+		keys = []
+		values = []
+		if self.thread_import_campaign.stopped():
+			return
+		for elements in self.campaign_info:
+			if elements.tag in ('id', 'landing_pages', 'messages', 'visits', 'credentials', 'deaddrop_deployments', 'deaddrop_connections'):
+				continue
+			keys.append(elements.tag)
+			values.append(elements.text)
+
+		self.rpc('db/table/set', 'campaigns', int(self.campaign_info.find('id').text), tuple(keys), tuple(values))
+		nodes_completed += float(len(values) + 1)
+		percentage_completed = nodes_completed / node_count
+		GLib.idle_add(self.import_progress.set_fraction, percentage_completed)
+		if self.thread_import_campaign.stopped():
+			return
+
+		for tables in ('landing_pages', 'messages', 'visits', 'credentials', 'deaddrop_deployments', 'deaddrop_connections'):
+			inserted_ids = []
+			if self.thread_import_campaign.stopped():
+				return
+			self._update_text_view('Serializing table {} data for import'.format(tables), idle=True)
+			keys, rows = self._get_keys_values(self.campaign_info.find(tables))
+			self._update_text_view('Working on table {} adding {} rows'.format(tables, len(rows)), idle=True)
+			if self.thread_import_campaign.stopped():
+				return
+
+			# make table rows easy to manage for updating new ids returned
+			table_rows = []
+			for row in rows:
+				row = dict(zip(keys, row))
+				table_rows.append(row)
+
+			while len(rows) > 0:
+				if self.thread_import_campaign.stopped():
+					return
+				try:
+					inserted_ids = inserted_ids + self.rpc('/db/table/insert/multi', tables, keys, rows[:batch_size], deconflict_ids=True)
+				except advancedhttpserver.RPCError:
+					response = gui_utilities.glib_idle_add_wait(self.failed_import_action)
+					if response == -8:
```

Comment:
There should either be a constant mapped to `-8` here or it needs a comment regarding it's significance.